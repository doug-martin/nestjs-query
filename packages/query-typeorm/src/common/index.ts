export * from './randomString';
export * from './typeorm';
