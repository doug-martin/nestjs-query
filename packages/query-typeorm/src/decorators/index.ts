export { InjectTypeOrmQueryService } from './inject-typeorm-query-service.decorator';
