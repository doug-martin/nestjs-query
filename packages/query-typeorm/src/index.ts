export { TypeOrmQueryService, TypeOrmQueryServiceOpts } from './services';
export { NestjsQueryTypeOrmModule } from './module';
