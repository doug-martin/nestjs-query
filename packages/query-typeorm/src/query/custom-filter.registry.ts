import { Class } from '@nestjs-query/core';
import { ColumnType, ObjectLiteral } from 'typeorm';

export type CustomFilterResult = { sql: string; params: ObjectLiteral };

export interface CustomFilterContext {
  fieldType?: ColumnType;
}

export interface CustomFilter<OperationType = string, Entity = unknown> {
  apply(
    field: keyof Entity | string,
    cmp: OperationType,
    val: unknown,
    alias?: string,
    context?: CustomFilterContext,
  ): CustomFilterResult;
}

type OperationCustomFilters = Record<string, CustomFilter>;

type SetFilterSpec =
  | { types: ColumnType[]; operations: string[] }
  | { klass: Class<unknown>; field: string; operations: string[] };

export class CustomFilterRegistry {
  // Registry for (class, field) filters
  private cfoRegistry: Map<Class<unknown>, Record<string | symbol | number, OperationCustomFilters>> = new Map();

  // Registry for (type) filters
  private tRegistry: Map<ColumnType, OperationCustomFilters> = new Map();

  getFilter<Entity = unknown>(
    operation: string,
    type?: ColumnType, // Type is optional, since we might have non property-backed filters
    klass?: Class<Entity>,
    field?: keyof Entity | string,
  ): CustomFilter<string, Entity> | undefined {
    // Most specific: (class, field) filters.
    if (klass && field) {
      const flt = this.cfoRegistry.get(klass)?.[field]?.[operation];
      if (flt) {
        return flt;
      }
    }
    // Type filters
    if (type) {
      return this.tRegistry.get(type)?.[operation];
    }
    return undefined;
  }

  /**
   * We have 2 types of filters:
   * - (type, operation) global filters
   * - (class, field, operation) filters
   * type is the database column type (TypeORM's ColumnType)
   * Specificity of the filters increases from top to bottom
   */
  setFilter<Entity>(filter: CustomFilter<string, Entity>, opts: SetFilterSpec): void {
    if ('klass' in opts && 'field' in opts) {
      const { klass, field, operations } = opts;
      if (!operations || operations.length === 0) {
        throw new Error('Cannot register a filter without operations, please define the operations array');
      }
      const entityFilters = this.cfoRegistry.get(klass) || {};
      entityFilters[field] = this.createCustomFilterOperationMap(filter, operations);
      this.cfoRegistry.set(klass, entityFilters);
    } else if ('types' in opts) {
      const { types, operations } = opts;
      if (!types || types.length === 0) {
        throw new Error('Cannot register a (type) filter without types, please define the types array');
      }
      if (!operations || operations.length === 0) {
        throw new Error('Cannot register a filter without operations, please define the operations array');
      }
      for (const type of types) {
        this.tRegistry.set(type, this.createCustomFilterOperationMap(filter, operations));
      }
    }
  }

  private createCustomFilterOperationMap(cf: CustomFilter, operations: string[]): OperationCustomFilters {
    const ocf: OperationCustomFilters = {};
    for (const op of operations) {
      ocf[op] = cf;
    }
    return ocf;
  }
}
