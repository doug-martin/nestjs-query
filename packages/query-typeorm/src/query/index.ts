export * from './filter-query.builder';
export * from './where.builder';
export * from './sql-comparison.builder';
export * from './relation-query.builder';
export * from './aggregate.builder';
