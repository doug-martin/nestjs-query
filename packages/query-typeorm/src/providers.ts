import { Class, getQueryServiceToken } from '@nestjs-query/core';
import { FactoryProvider } from '@nestjs/common';
import { getRepositoryToken } from '@nestjs/typeorm';
import { Repository, Connection, ConnectionOptions } from 'typeorm';
import { TypeOrmQueryService } from './services';

function createTypeOrmQueryServiceProvider<Entity>(
  EntityClass: Class<Entity>,
  connection?: Connection | ConnectionOptions | string,
): FactoryProvider {
  return {
    provide: getQueryServiceToken(EntityClass),
    useFactory(repo: Repository<Entity>) {
      return new TypeOrmQueryService(repo);
    },
    inject: [getRepositoryToken(EntityClass, connection)],
  };
}

export const createTypeOrmQueryServiceProviders = (
  entities: Class<unknown>[],
  connection?: Connection | ConnectionOptions | string,
): FactoryProvider[] => entities.map((entity) => createTypeOrmQueryServiceProvider(entity, connection));
