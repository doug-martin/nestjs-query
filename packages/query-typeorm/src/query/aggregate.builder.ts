import { SelectQueryBuilder } from 'typeorm';
import { AggregateQuery, AggregateResponse } from '@nestjs-query/core';
import { BadRequestException } from '@nestjs/common';
import { camelCase } from 'camel-case';

enum AggregateFuncs {
  AVG = 'AVG',
  SUM = 'SUM',
  COUNT = 'COUNT',
  MAX = 'MAX',
  MIN = 'MIN',
}

const AGG_REGEXP = /(AVG|SUM|COUNT|MAX|MIN|GROUP_BY)_(.*)/;

/**
 * @internal
 * Builds a WHERE clause from a Filter.
 */
export class AggregateBuilder<Entity> {
  // eslint-disable-next-line @typescript-eslint/no-shadow
  static async asyncConvertToAggregateResponse<Entity>(
    responsePromise: Promise<Record<string, unknown>[]>,
  ): Promise<AggregateResponse<Entity>[]> {
    const aggResponse = await responsePromise;
    return this.convertToAggregateResponse(aggResponse);
  }

  // eslint-disable-next-line @typescript-eslint/no-shadow
  static getAggregateSelects<Entity>(query: AggregateQuery<Entity>): string[] {
    return [...this.getAggregateGroupBySelects(query), ...this.getAggregateFuncSelects(query)];
  }

  // eslint-disable-next-line @typescript-eslint/no-shadow
  private static getAggregateGroupBySelects<Entity>(query: AggregateQuery<Entity>): string[] {
    return (query.groupBy ?? []).map((f) => this.getGroupByAlias(f));
  }

  // eslint-disable-next-line @typescript-eslint/no-shadow
  private static getAggregateFuncSelects<Entity>(query: AggregateQuery<Entity>): string[] {
    const aggs: [AggregateFuncs, (keyof Entity)[] | undefined][] = [
      [AggregateFuncs.COUNT, query.count],
      [AggregateFuncs.SUM, query.sum],
      [AggregateFuncs.AVG, query.avg],
      [AggregateFuncs.MAX, query.max],
      [AggregateFuncs.MIN, query.min],
    ];
    return aggs.reduce((cols, [func, fields]) => {
      const aliases = (fields ?? []).map((f) => this.getAggregateAlias(func, f));
      return [...cols, ...aliases];
    }, [] as string[]);
  }

  // eslint-disable-next-line @typescript-eslint/no-shadow
  static getAggregateAlias<Entity>(func: AggregateFuncs, field: keyof Entity): string {
    return `${func}_${field as string}`;
  }

  // eslint-disable-next-line @typescript-eslint/no-shadow
  static getGroupByAlias<Entity>(field: keyof Entity): string {
    return `GROUP_BY_${field as string}`;
  }

  // eslint-disable-next-line @typescript-eslint/no-shadow
  static convertToAggregateResponse<Entity>(rawAggregates: Record<string, unknown>[]): AggregateResponse<Entity>[] {
    return rawAggregates.map((response) => {
      return Object.keys(response).reduce((agg: AggregateResponse<Entity>, resultField: string) => {
        const matchResult = AGG_REGEXP.exec(resultField);
        if (!matchResult) {
          throw new Error('Unknown aggregate column encountered.');
        }
        const [matchedFunc, matchedFieldName] = matchResult.slice(1);
        const aggFunc = camelCase(matchedFunc.toLowerCase()) as keyof AggregateResponse<Entity>;
        const fieldName = matchedFieldName as keyof Entity;
        const aggResult = agg[aggFunc] || {};
        return {
          ...agg,
          // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
          [aggFunc]: { ...aggResult, [fieldName]: response[resultField] },
        };
      }, {});
    });
  }

  /**
   * Builds a aggregate SELECT clause from a aggregate.
   * @param qb - the `typeorm` SelectQueryBuilder
   * @param aggregate - the aggregates to select.
   * @param alias - optional alias to use to qualify an identifier
   */
  build<Qb extends SelectQueryBuilder<Entity>>(qb: Qb, aggregate: AggregateQuery<Entity>, alias?: string): Qb {
    const selects = [
      ...this.createGroupBySelect(aggregate.groupBy, alias),
      ...this.createAggSelect(AggregateFuncs.COUNT, aggregate.count, alias),
      ...this.createAggSelect(AggregateFuncs.SUM, aggregate.sum, alias),
      ...this.createAggSelect(AggregateFuncs.AVG, aggregate.avg, alias),
      ...this.createAggSelect(AggregateFuncs.MAX, aggregate.max, alias),
      ...this.createAggSelect(AggregateFuncs.MIN, aggregate.min, alias),
    ];
    if (!selects.length) {
      throw new BadRequestException('No aggregate fields found.');
    }
    const [head, ...tail] = selects;
    return tail.reduce(
      (acc: Qb, [select, selectAlias]) => acc.addSelect(select, selectAlias),
      qb.select(head[0], head[1]),
    );
  }

  private createAggSelect(func: AggregateFuncs, fields?: (keyof Entity)[], alias?: string): [string, string][] {
    if (!fields) {
      return [];
    }
    return fields.map((field) => {
      const col = alias ? `${alias}.${field as string}` : (field as string);
      return [`${func}(${col})`, AggregateBuilder.getAggregateAlias(func, field)];
    });
  }

  private createGroupBySelect(fields?: (keyof Entity)[], alias?: string): [string, string][] {
    if (!fields) {
      return [];
    }
    return fields.map((field) => {
      const col = alias ? `${alias}.${field as string}` : (field as string);
      return [`${col}`, AggregateBuilder.getGroupByAlias(field)];
    });
  }
}
