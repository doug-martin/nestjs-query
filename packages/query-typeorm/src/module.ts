import { Class } from '@nestjs-query/core';
import { TypeOrmModule } from '@nestjs/typeorm';
import { DynamicModule } from '@nestjs/common';
import { Connection, ConnectionOptions } from 'typeorm';
import { createTypeOrmQueryServiceProviders } from './providers';

export class NestjsQueryTypeOrmModule {
  static forFeature(entities: Class<unknown>[], connection?: Connection | ConnectionOptions | string): DynamicModule {
    const queryServiceProviders = createTypeOrmQueryServiceProviders(entities, connection);
    const typeOrmModule = TypeOrmModule.forFeature(entities, connection);
    return {
      imports: [typeOrmModule],
      module: NestjsQueryTypeOrmModule,
      providers: [...queryServiceProviders],
      exports: [...queryServiceProviders, typeOrmModule],
    };
  }
}
