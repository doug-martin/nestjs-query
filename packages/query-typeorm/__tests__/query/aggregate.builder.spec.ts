/* eslint-disable @typescript-eslint/naming-convention */
import { AggregateQuery } from '@nestjs-query/core';
import { closeTestConnection, createTestConnection, getTestConnection } from '../__fixtures__/connection.fixture';
import { TestEntity } from '../__fixtures__/test.entity';
import { AggregateBuilder } from '../../src/query';

describe('AggregateBuilder', (): void => {
  beforeEach(createTestConnection);
  afterEach(closeTestConnection);

  const getRepo = () => getTestConnection().getRepository(TestEntity);
  const getQueryBuilder = () => getRepo().createQueryBuilder();
  const createAggregateBuilder = () => new AggregateBuilder<TestEntity>();

  const expectSQLSnapshot = (agg: AggregateQuery<TestEntity>): void => {
    const selectQueryBuilder = createAggregateBuilder().build(getQueryBuilder(), agg, 'TestEntity');
    const [sql, params] = selectQueryBuilder.getQueryAndParameters();
    expect(sql).toMatchSnapshot();
    expect(params).toMatchSnapshot();
  };

  it('should throw an error if no selects are generated', (): void => {
    expect(() => createAggregateBuilder().build(getQueryBuilder(), {})).toThrow('No aggregate fields found.');
  });

  it('should create selects for all aggregate functions', (): void => {
    expectSQLSnapshot({
      count: ['testEntityPk'],
      avg: ['numberType'],
      sum: ['numberType'],
      max: ['stringType', 'dateType', 'numberType'],
      min: ['stringType', 'dateType', 'numberType'],
    });
  });

  it('should create selects for all aggregate functions and group bys', (): void => {
    expectSQLSnapshot({
      groupBy: ['stringType', 'boolType'],
      count: ['testEntityPk'],
    });
  });

  describe('.convertToAggregateResponse', () => {
    it('should convert a flat response into an Aggregate response', () => {
      const dbResult = [
        {
          GROUP_BY_stringType: 'z',
          COUNT_testEntityPk: 10,
          SUM_numberType: 55,
          AVG_numberType: 5,
          MAX_stringType: 'z',
          MAX_numberType: 10,
          MIN_stringType: 'a',
          MIN_numberType: 1,
        },
      ];
      expect(AggregateBuilder.convertToAggregateResponse<TestEntity>(dbResult)).toEqual([
        {
          groupBy: { stringType: 'z' },
          count: { testEntityPk: 10 },
          sum: { numberType: 55 },
          avg: { numberType: 5 },
          max: { stringType: 'z', numberType: 10 },
          min: { stringType: 'a', numberType: 1 },
        },
      ]);
    });

    it('should throw an error if a column is not expected', () => {
      const dbResult = [
        {
          COUNTtestEntityPk: 10,
        },
      ];
      expect(() => AggregateBuilder.convertToAggregateResponse<TestEntity>(dbResult)).toThrow(
        'Unknown aggregate column encountered.',
      );
    });
  });
});
