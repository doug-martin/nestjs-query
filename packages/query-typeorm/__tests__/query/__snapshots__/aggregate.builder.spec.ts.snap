// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AggregateBuilder should create selects for all aggregate functions 1`] = `SELECT COUNT("TestEntity"."test_entity_pk") AS "COUNT_testEntityPk", SUM("TestEntity"."number_type") AS "SUM_numberType", AVG("TestEntity"."number_type") AS "AVG_numberType", MAX("TestEntity"."string_type") AS "MAX_stringType", MAX("TestEntity"."date_type") AS "MAX_dateType", MAX("TestEntity"."number_type") AS "MAX_numberType", MIN("TestEntity"."string_type") AS "MIN_stringType", MIN("TestEntity"."date_type") AS "MIN_dateType", MIN("TestEntity"."number_type") AS "MIN_numberType" FROM "test_entity" "TestEntity"`;

exports[`AggregateBuilder should create selects for all aggregate functions 2`] = `Array []`;

exports[`AggregateBuilder should create selects for all aggregate functions and group bys 1`] = `SELECT "TestEntity"."string_type" AS "GROUP_BY_stringType", "TestEntity"."bool_type" AS "GROUP_BY_boolType", COUNT("TestEntity"."test_entity_pk") AS "COUNT_testEntityPk" FROM "test_entity" "TestEntity"`;

exports[`AggregateBuilder should create selects for all aggregate functions and group bys 2`] = `Array []`;
